File: DrawDemoSwift/Old ObjC/GCSPolarDuplicateController.m
#import "GCSPolarDuplicateController.h"

#import <DKDrawKit/LogEvent.h>

@implementation GCSPolarDuplicateController
#pragma mark As a GCPolarDuplicateController

- (IBAction)angleAction:(id)sender
{
#pragma unused(sender)
}

- (IBAction)cancelAction:(id)sender
{
#pragma unused(sender)
	[self.window orderOut:self];
	[NSApp endSheet:self.window returnCode:NSCancelButton];
}

- (IBAction)centreAction:(id)sender
{
#pragma unused(sender)
	// disable the OK button if either of the centre fields are empty
	[self conditionallyEnableOKButton];
}

- (IBAction)copiesAction:(id)sender
{
#pragma unused(sender)
}

- (IBAction)duplicateAction:(id)sender
{
#pragma unused(sender)
	[self.window orderOut:self];
	[NSApp endSheet:self.window returnCode:NSOKButton];
}

- (IBAction)rotateCopiesAction:(id)sender
{
#pragma unused(sender)
}

- (IBAction)autoFitAction:(id)sender
{
	BOOL enable = ([sender intValue] == 0);

	//[mManualSettingsBox setEnabled:enable];

	mAngleIncrementTextField.enabled = enable;
	mCopiesTextField.enabled = enable;
	mRotateCopiesCheckbox.intValue = 1;
	mRotateCopiesCheckbox.enabled = enable;
}

#pragma mark -
- (void)beginPolarDuplicationDialog:(NSWindow *)parentWindow polarDelegate:(id)delegate
{
	mDelegateRef = delegate;

	[NSApp beginSheet:self.window
		modalForWindow:parentWindow
		 modalDelegate:self
		didEndSelector:@selector(sheetDidEnd:returnCode:contextInfo:)
		   contextInfo:@"polar_duplication"];

	NSInteger items = [delegate countOfItemsInSelection];
	mAutoFitCircleCheckbox.enabled = (items == 1);
	[self conditionallyEnableOKButton];
}

- (void)sheetDidEnd:(NSWindow *)sheet returnCode:(NSInteger)returnCode contextInfo:(void *)contextInfo
{
#pragma unused(sheet, contextInfo)
	if (returnCode == NSOKButton) {
		// extract parameters and do something with them

		NSInteger copies = mCopiesTextField.integerValue;
		NSPoint centre;

		centre.x = mCentreXTextField.floatValue;
		centre.y = mCentreYTextField.floatValue;

		CGFloat incAngle = mAngleIncrementTextField.floatValue;
		BOOL rotCopies = mRotateCopiesCheckbox.intValue;

		if (mAutoFitCircleCheckbox.intValue == 1) {
			[mDelegateRef doAutoPolarDuplicateWithCentre:centre];
		} else {

			LogEvent_(kReactiveEvent, @"dialog data: copies %ld; centre {%.2f,%.2f}; incAngle %.3f; rotateCopies %d", (long)copies, centre.x, centre.y, incAngle, rotCopies);

			[mDelegateRef doPolarDuplicateCopies:copies centre:centre incAngle:incAngle rotateCopies:rotCopies];
		}
	}
}

#pragma mark -
- (void)conditionallyEnableOKButton
{
	if (mCentreXTextField.stringValue == nil ||
		[mCentreXTextField.stringValue isEqualToString:@""] ||
		mCentreYTextField.stringValue == nil ||
		[mCentreYTextField.stringValue isEqualToString:@""])
		[mOKButton setEnabled:NO];
	else
		[mOKButton setEnabled:YES];
}

@end

File: DrawDemoSwift/GCSPolarDuplicateController.swift
//
//  GCSPolarDuplicateController.swift
//  DrawDemoSwift
//
//  Created by C.W. Betts on 1/25/18.
//  Copyright © 2018 DrawKit. All rights reserved.
//

import Foundation

protocol GCSPolarDuplicationDelegate: NSObjectProtocol {
	func doPolarDuplicateCopies(_ copies: Int, centre cp: NSPoint, incAngle angle: CGFloat, rotateCopies rotCopies: Bool)
	func doAutoPolarDuplicate(withCentre cp: NSPoint)
	var countOfItemsInSelection: Int { get }
}

class GCSPolarDuplicateController: NSWindowController {
	@IBOutlet weak var angleIncrementTextField: NSTextField!
	@IBOutlet weak var centreXTextField: NSTextField!
	@IBOutlet weak var centreYTextField: NSTextField!
	@IBOutlet weak var copiesTextField: NSTextField!
	@IBOutlet weak var rotateCopiesCheckbox: NSButton!
	@IBOutlet weak var autoFitCircleCheckbox: NSButton!
	@IBOutlet weak var okButton: NSButton!
	@IBOutlet weak var cancelButton: NSButton!
	@IBOutlet weak var manualSettingsBox: NSBox!
	
	private weak var parentWindow: NSWindow?
	
	@IBAction open func angleAction(_ sender: Any?) {
		// empty
	}
	
	@IBAction open func cancelAction(_ sender: Any?) {
		window?.orderOut(self)
		parentWindow?.endSheet(window!, returnCode: .cancel)
	}
	
	@IBAction open func centreAction(_ sender: Any?) {
		// disable the OK button if either of the centre fields are empty
		conditionallyEnableOKButton()
	}
	
	@IBAction open func copiesAction(_ sender: Any?) {
		// empty
	}
	
	@IBAction open func duplicateAction(_ sender: Any?) {
		window?.orderOut(self)
		parentWindow?.endSheet(window!, returnCode: .OK)
	}
	
	@IBAction open func rotateCopiesAction(_ sender: Any?) {
		// empty
	}
	
	@IBAction open func autoFitAction(_ sender: Any?) {
		let enable: Bool = {
			guard let state: NSControl.StateValue = (sender as AnyObject?)?.state else {
				return false
			}
			return state == .on
		}()
		
		//[mManualSettingsBox setEnabled:enable];

		angleIncrementTextField.isEnabled = enable
		copiesTextField.isEnabled = enable
		rotateCopiesCheckbox.state = .on
		rotateCopiesCheckbox.isEnabled = enable
	}

	func beginPolarDuplicationDialog(_ parentWindow: NSWindow, polarDelegate delegate: GCSPolarDuplicationDelegate) {
		self.parentWindow = parentWindow
		
		parentWindow.beginSheet(self.parentWindow!) { (returnCode) in
			if returnCode == .OK {
				// extract parameters and do something with them
				let copies = self.copiesTextField.integerValue
				
				let centre = NSPoint(x: self.centreXTextField.objectValue as? CGFloat ?? 0, y: self.centreYTextField.objectValue as? CGFloat ?? 0)
				
				let incAngle = self.angleIncrementTextField.objectValue as? CGFloat ?? 0
				let rotCopies = self.rotateCopiesCheckbox.state == .on
				
				if self.autoFitCircleCheckbox.state == .on {
					delegate.doAutoPolarDuplicate(withCentre: centre)
				} else {
					LogEvent(.reactiveEvent, String(format: "dialog data: copies %ld; centre {%.2f,%.2f}; incAngle %.3f; rotateCopies %d", copies, centre.x, centre.y, rotCopies ? 1 : 0))
					
					delegate.doPolarDuplicateCopies(copies, centre: centre, incAngle: incAngle, rotateCopies: rotCopies)
				}
			}
			self.parentWindow = nil
		}
		
		let items = delegate.countOfItemsInSelection
		autoFitCircleCheckbox.isEnabled = items == 1
		conditionallyEnableOKButton()
	}
	
	private func conditionallyEnableOKButton() {
		if centreXTextField.stringValue == "" || centreYTextField.stringValue == "" {
			okButton.isEnabled = false
		} else {
			okButton.isEnabled = true
		}
	}
}
